ENH: spatial.cKDTree: Add multi-thread build

jkelling opened the pull request:
This PR adds threading to the build phase `KDTree`/`cKDTree`. With large number of points (>~100M) the single-core build time grows into tens of minutes. Considering the memory requirements at this scale, this also means lots of cores sitting idle.

#### Reference issue
N/A

#### What does this implement/fix?
`cKDTree`'s `build()` function is changed to spawn worker threads at each recursion level up to a maximum number of workers controlled by the user through the optional `build_workers` argument to `cKDTree.__init__()`.
* The parallelism is implemented such, that the resulting tree is identical to the one from previous sequential implementation.
* The code uses C++11 threads. (I did not find any documentation or examples in the source as to what threading model is preferred in SciPy C++ code,so I chose the standard way. Should be supported even by GCC 4.8)
* There is no dynamic load balancing to keep the code simple. Dynamic balancing would probably not improve performance enough to be worth the added complexity.
* Performance limitations: Maximum speedup on Zen2 and SkyLake is between 4 and 5x with diminishing returns above ~four threads:
  * mainly due to memory bandwidth limitations
  * load imbalances
  * the top level sort has to done by one thread; C++17 parallel STL would help out here.

#### Additional information
I did no find a C++ style guide for SciPy, so I tried to follow the style in the present code.

tupui commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -415,7 +416,7 @@ cdef np.intp_t get_num_workers(workers: object, kwargs: dict) except -1:
     """
     cKDTree(data, leafsize=16, compact_nodes=True, copy_data=False,
-            balanced_tree=True, boxsize=None)
+            balanced_tree=True, boxsize=None, build_workers=1)
```
With new arguments, we make them keyword only arguments
```suggestion
            balanced_tree=True, boxsize=None, *, workers=1)
```

tupui commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -459,6 +460,9 @@ cdef class cKDTree:
         is the boxsize along i-th dimension. The input data shall be wrapped
         into :math:`[0, L_i)`. A ValueError is raised if any of the data is
         outside of this bound.
+    build_workers : int, optional
+        Number of workers to use for parallel processing. If <=1 one thread is
+        used. Default: 1.
```
In other places, we use `workers` and if set to -1, it adjusts to the machine max number of workers. Same in the other functions.
```suggestion
    workers : int, optional
        Number of workers to use for parallel processing. If -1 is given all
        CPU threads are used. Default is 1.
```

jkelling replied:
This is not the currently implemented behavior, but I can add it to the commit....

tupui replied:
The -1 you mean? IMHO, I think this would be better if you could. We would have something more coherent with the rest. Thanks!

tupui commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -548,14 +552,14 @@ cdef class cKDTree:
 
     def __init__(cKDTree self, data, np.intp_t leafsize=16, compact_nodes=True,
-            copy_data=False, balanced_tree=True, boxsize=None):
+            copy_data=False, balanced_tree=True, boxsize=None, build_workers=1):
```
```suggestion
            copy_data=False, balanced_tree=True, boxsize=None, *, workers=1):
```

tupui reviewed (changes requested):
Thanks @jkelling! I will let others comment on the cython, I just have a few suggestions.

jkelling commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -610,7 +615,7 @@ cdef class cKDTree:
         ptmpmins = &tmpmins[0]
         with nogil:
-            build_ckdtree(cself, 0, cself.n, ptmpmaxes, ptmpmins, median, compact)
+            build_ckdtree(cself, 0, cself.n, ptmpmaxes, ptmpmins, median, compact, workers)
```
```suggestion
            build_ckdtree(cself, 0, cself.n, ptmpmaxes, ptmpmins, median, compact, build_workers)
```

jkelling commented on scipy/spatial/kdtree.py:
```diff
@@ -334,14 +337,14 @@ def tree(self):
 
     def __init__(self, data, leafsize=10, compact_nodes=True, copy_data=False,
-                 balanced_tree=True, boxsize=None):
+                 balanced_tree=True, boxsize=None, build_workers=1):
```
```suggestion
                 balanced_tree=True, boxsize=None, *, workers=1):
```

jkelling commented on scipy/spatial/kdtree.py:
```diff
@@ -334,14 +337,14 @@ def tree(self):
         # Note KDTree has different default leafsize from cKDTree
         super().__init__(data, leafsize, compact_nodes, copy_data,
-                         balanced_tree, boxsize)
+                         balanced_tree, boxsize, build_workers)
```
```suggestion
                         balanced_tree, boxsize, workers)
```

tupui commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -601,6 +605,7 @@ cdef class cKDTree:
 
         compact = 1 if compact_nodes else 0
         median = 1 if balanced_tree else 0
+        workers = build_workers
```
We have this logic someplace else. I think this would be better to just copy this (alternatively make it as a util function to be reused elsewhere).
```python
    workers = int(workers)
    if workers == -1:
        workers = os.cpu_count()  # type: ignore[assignment]
        if workers is None:
            raise NotImplementedError(
                "Cannot determine the number of cpus using os.cpu_count(), "
                "cannot use -1 for the number of workers"
            )
    elif workers <= 0:
        raise ValueError(f"Invalid number of workers: {workers}, must be -1 "
                         "or > 0")
```

jkelling replied:
I did actually copy from here and adapted it, just missed the detail that `os.cpu_count()` may return `None`. I edited my suggestion to fix.

tupui reviewed (commented):
The changes seems fine, feel free to batch the suggestions so that it just do a single commit.

jkelling commented:
> The changes seems fine, feel free to batch the suggestions so that it just do a single commit.

Done (with an adjustment to follow your other comment). Thanks.

tylerjereddy commented on scipy/spatial/kdtree.py:
```diff
@@ -222,6 +222,9 @@ class KDTree(cKDTree):
         is the boxsize along i-th dimension. The input data shall be wrapped
         into :math:`[0, L_i)`. A ValueError is raised if any of the data is
         outside of this bound.
+    build_workers : int, optional
```
We better check with @peterbell10, @rainwoodman, and @sturlamolden about the API here.

Plus, they are each capable of reviewing C++ changes.

Another question might be related to `asv` benchmarks/testing with > 1 thread, but probably best to see what they say initially.

tupui replied:
Just to be clear. On my side I am just commenting on the interface to make it coherent with such existing options.

jkelling replied:
> Another question might be related to asv benchmarks/unit testing with > 1 thread, but probably best to see what they say initially.

I agree with this,. but I do not know how to add such tests. Do you have any pointers?

jkelling replied:
I added a test dimensionwith `[1, 2]` workers [here](https://github.com/scipy/scipy/pull/14481/files#diff-5c9e7e8948e7ff6efc82ab31a7edc2f9cc09abcd130808520be342f4afa76668)

sturlamolden commented:
I am not familiar with C++11 async APIs (I’m too old, and used to OpenMP and GCD). However, it is clear that the build phase can be made parallel with fork-join logic. Also I would like to request that there be zero overhead (speed and memory) for the single-threaded case. 

Another thing I would like to comment relates to portability. In the query functions we have used Python threading.Thread to make sure cKDTree is portable to the same platforms as Python. Is there any reason we should avoid Python threading here? How is the portability of C++11 async compared to Python?

jkelling commented:
> I am not familiar with C++11 async APIs (I’m too old, and used to OpenMP and GCD). However, it is clear that the build phase can be made parallel with fork-join logic. Also I would like to request that there be zero overhead (speed and memory) for the single-threaded case.

There is currently a slight overhead in runtime inf the single threaded cased case compared to master which is because the same logic as in the multi-threaded case is used to construct `tree_buffer`, while on master the same stem is performed inside the recursion with better cache-locality. I will check if I can remove this overhead with minimal extra logic.

In some cases, having no speed overhead at all for the single thread case compared to a sequential implementation basically requires the single-thread case to be handled separately. How much complexity is this point worth to you?
 
> Another thing I would like to comment relates to portability. In the query functions we have used Python threading.Thread to make sure cKDTree is portable to the same platforms as Python. Is there any reason we should avoid Python threading here? How is the portability of C++11 async compared to Python?

I assume, that to use Python threads, the recursion would need to be implemented in python. This alone would create overhead for the sequential case. Note, that `query` is already threaded using Python threads, which possible because the problem is embarrassingly parallel, which each Python thread just calling the C++ `query` implementation. C++ threads, in the way used here, are supported by any compiler that supports C++11, as they are part of the STL.

tupui commented on scipy/spatial/kdtree.py:
```diff
@@ -222,6 +222,9 @@ class KDTree(cKDTree):
         is the boxsize along i-th dimension. The input data shall be wrapped
         into :math:`[0, L_i)`. A ValueError is raised if any of the data is
         outside of this bound.
+    build_workers : int, optional
```
Just to be clear. On my side I am just commenting on the interface to make it coherent with such existing options.

The doc should be adapted as well.
```suggestion
    workers : int, optional
```

tupui commented on scipy/spatial/kdtree.py:
```diff
@@ -222,6 +222,9 @@ class KDTree(cKDTree):
         is the boxsize along i-th dimension. The input data shall be wrapped
         into :math:`[0, L_i)`. A ValueError is raised if any of the data is
         outside of this bound.
+    build_workers : int, optional
+        Number of workers to use for parallel processing. If <=1 one thread is
+        used. Default: 1.
```
```suggestion
    workers : int, optional
        Number of workers to use for parallel processing. If -1 is given all
        CPU threads are used. Default is 1.
```

jkelling replied:
missed this one. I will push it with some later changes.

jkelling replied:
Amended to https://github.com/scipy/scipy/pull/14481/commits/5e74b0a3f67801bd7f67550eed4e397fc4fbb351 .

jkelling commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -153,51 +195,97 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+    }
+
+    // merge sections from buflist in self->tree_buffer and update pointers and indices
+    std::deque<std::future<void>> futures;
```
I will rather change this to call `futures.reserve(_workers)`, so that in the following loop I do not need to track an index and can used `push_back()`.

jkelling replied:
Addressed in https://github.com/scipy/scipy/commit/fbb4a994ae81441c09baa299d8f9b0c1ce54f4ef .

jkelling commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -153,51 +195,97 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+    std::deque<std::future<void>> futures;
+    ckdtree_intp_t offset = 0;
+    for(auto& b : buflist) {
+        futures.push_back(std::async(std::launch::async, [offset, &b, self]() -> void {
```
I should split the loop and have the master thread handle the last element of `buflist` itself, which eliminates an unnecessary  `asynch` call in the `_workers==1` case.

jkelling replied:
Adressed in https://github.com/scipy/scipy/commit/fbb4a994ae81441c09baa299d8f9b0c1ce54f4ef .

jkelling commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -153,51 +195,97 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
         if (CKDTREE_LIKELY(_compact)) {
-            _less = build(self, start_idx, p, maxes, mins, _median, _compact);
-            _greater = build(self, p, end_idx, maxes, mins, _median, _compact);
+            workers -= worker.run(self, p, end_idx, maxes, mins, _median, _compact, workers);
+            _less = build(self, start_idx, p, maxes, mins, _median, _compact, buflist, workers);
+            if(!worker)
+                _greater = build(self, p, end_idx, maxes, mins, _median, _compact, buflist, workers);
```
I did think about `reserve` and tried it: The effect is not noticable. `std::vector` scales up the allocation sizes each time it has to resize, such that for large trees it does even become a bit faster than `deque`.

In general, I think `reserve` would be a good idea, because on some systems it may provide an advantage. I was reluctant to add `reserve` because it was not there before and had the same thought as @sturlamolden  about the memory overhead: in an imbalanced tree, using `reserve` would allocate more memory than the sequential implementation.

jkelling replied:
> How about instead of modifying this to handle both single and multithreaded, you could move the partitioning code into a helper function and have two versions of build.

If the PR would be accepted this way, I could do that. I was going for a minimal patch size in my implementation. The version with one `build` function would also be more compatible with an approach with dynamic load balancing, because the the fanning-out would not be static.

jkelling replied:
The reserve is in https://github.com/scipy/scipy/commit/fbb4a994ae81441c09baa299d8f9b0c1ce54f4ef . I am still concerned, that the two build functions would duplicate too much code.

peterbell10 replied:
>I am still concerned, that the two build functions would duplicate too much code.

This is what I expect `build_parallel` would look like:
```cpp
static ckdtree_intp_t
build_parallel(
      ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
      double *maxes, double *mins, const int _median, const int _compact
      std::list<std::vector> *buflist, int workers)
{
    auto &node_buffer = buflist->back();
    const int min_parallel_leafs = /* Some minimum grain size */
    if (end_idx-start_idx <= self->leafsize * min_parallel_leafs) {
        return build(..., node_buffer);
    }

    auto res = split_and_partition(start_idx, end_idx, maxes, mins, _median, _compact);
    ckdtree_intp_t d = res.d;
    ckdtree_intp_t p = res.p;
    double split = res.split;

    auto node_index = allocate_node(node_buffer, start_idx, end_idx);
    
    if (CKDTREE_UNLIKELY(d == -1)) {
        // return leaf node
        ckdtree* n = &node_buffer[node_index];
        n->split_dim = -1;
        return node_index;
    }

    if (CKDTREE_LIKELY(_compact)) {
        /* launch one of these in parallel */
        build_worker worker(...);
        _greater = build_parallel(...);
        _less = worker.future.get();
    }
    else
    {
        std::vector<double> mids(maxes, maxes + m);
        mids[d] = split;
        build_worker worker(...);

        for (i=0; i<m; ++i) mids[i] = mins[i];
        mids[d] = split;
        _greater = build(self, p, end_idx, maxes, mids, _median, _compact);
        _less = worker.future.get();
    }

    
    /* fill in entries */
    ckdtreenode *n = &node_buffer[node_index];
    n->_less = _less;
    n->_greater = _greater;
    n->split_dim = d;
    n->split = split;

    return node_index;
}
```

Yes there is necessarily duplicates, but most of the heavy lifting has been factored into `split_and_partition` which determines the split point and partitions the indices.

jkelling replied:
This case cannot be factored-out:
https://github.com/scipy/scipy/blob/cd4029534163a43d2b6679578c5f60322e88a002/scipy/spatial/ckdtree/src/build.cxx#L203-L204
as-is, because it calls `build`. The correct `build` would need to passed as a pointer or a template predicate, at lease the first one would require `split_and_partition` the `buflist` and `workers` arguments.

Is there anything speaking against having a loop instead of the recursive call?
```c++
do {
 // partitioning
 //....

 if (CKDTREE_UNLIKELY(p == start_idx || p == end_idx)) {
   assert(!_compact);
   const auto fixed_val = data[indices[start_idx]*m + d];
   mins[d] = fixed_val; // tmp_mins is not needed as mins and maxs are modified in the  _compact case anyway,
   maxes[d] = fixed_val; // i.e. they are temporaies
 }
} while(CKDTREE_UNLIKELY(p == start_idx || p == end_idx));
```

jkelling replied:
I pushed a split version using the loop. I took the liberty of also cleaning up the local variable declarations (declare on use, not Fortran-Style; and more const).

jkelling commented:
I pushed a new version with some [changes](https://github.com/scipy/scipy/commit/fbb4a994ae81441c09baa299d8f9b0c1ce54f4ef) to address points raised by @peterbell10 and @sturlamolden .

On thing, is that I went back to `std::vector` instead of `std::deque` so that the copy of the temporary tree buffer(s) into `tree_buffer` can be replaced by a move in the case `workers_ == 1`. This has a slight negative impact on performance compared to the previous version if the vectors are not pre-allocated using `std::vector::reserve` (5~8% with ~3e8 points, but run-to-run variance tends to be not much less than this (I have no resources for more precise measurements ATM.)). Calls to `reserve` are currently in the code and active, but with preprocessor guards around them. Please tell me, if you want them in or out. Not sure if there is unanimity about this in the thread.

sturlamolden commented on scipy/spatial/kdtree.py:
```diff
@@ -222,6 +222,9 @@ class KDTree(cKDTree):
         into :math:`[0, L_i)`. A ValueError is raised if any of the data is
         outside of this bound.
+    workers : int, optional
+        Number of workers to use for parallel processing. If -1 is given all
```
It should say this is for parallel build, so the users do not assume the queries will be parallel as well.

jkelling replied:
Changed to "[...] for parallel build."

sturlamolden commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -153,51 +200,120 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
-            _less = build(self, start_idx, p, maxes, mins, _median, _compact);
-            _greater = build(self, p, end_idx, maxes, mins, _median, _compact);
+            if(workers > 1) {
+                worker.reset(new build_worker());
```
RAII violation

jkelling replied:
Right, I will fix it after the discussions.

The constructor and the run function are still split because I am waiting for some reviews on this:
Should I:
1. Put the `build_worker` instance on the stack, as it was before, (no pointer, need `run()`) or
2. Have a heap allocation to reduce the stack size for the recursion (pointer, c'tor and `run()` are the same)?

I do not see a significant performance difference. In the interest of lower memory overhead of the patch, option 2 would be preferred.

I could also move the `unique_ptr` into `build_worker`, having kind-of a PIMPL. This would make option 2 look like option 1 in `build()`, but come with a nested class. I am still sure which style is preferred. The current code looks too much like C for me to assume that this pattern would be welcomed.

Please choose one option.

jkelling replied:
I added version three, please have a look.

peterbell10 replied:
If you implement https://github.com/scipy/scipy/pull/14481#discussion_r679274205 then exactly `workers` instances of `build_worker` will be created. At most `ceil(log2(workers))` on a single thread. Therefore creating it on the stack is not an issue.

jkelling replied:
There is another small reason to have it on the heap: The member `buflist_` of  `build_worker` is modified by the child thread, this would lead to false sharing of cache lines. It is actually only touched by the child trice though: when creating the initial buffer and retrieving it and when splicing-in the `buflist_` of the grandchild (the read when iterating before splicing will not cached anyway).

peterbell10 replied:
Just use `alignas(64)` on the stack variable and it will be on its own cache line.

jkelling replied:
This is not very portable, as it depends on the CPU (e.g. Power9 has 128 byte cache lines), so one would actually need to determine the cache line size of the target CPU, for which I do not know a portable way.

I put it in anyway as it makes the code a bit shorter and the false sharing is not a bog issue due to the infrequent access.

jkelling commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -12,29 +12,76 @@
+
+            future_ = std::async(std::launch::async, [&]() {
+                    buflist_.emplace_back();
+#ifdef STD_VECTOR_RESERVE
+                    buflist_.back().reserve((end_idx-start_idx)/self->leafsize/subworkers);
+#endif
```
I am not actually suggesting to leave this conditional compilation in: Should we `.reserve()` or not? @sturlamolden  @peterbell10

peterbell10 replied:
My vote is for reserve.

I think the worst case scenario is an unbalanced tree where half the threads threads immediately terminate with a leaf node and the other half holds almost all of the inner nodes. Even in this worst-case, I think the memory usage would only be ~1.5x that of a balanced tree. I don't think that's too unreasonable. Especially as it's only a temporary spike in allocated memory.

jkelling replied:
Removed the guards and left the reserve.

jkelling commented:
Is the Windows Azure runs timing-out a genera problem or could it point to a problem with this PR?

peterbell10 commented:
>Is the Windows Azure runs timing-out a genera problem or could it point to a problem with this PR?

Windows CI is a bit flaky unforunately. Timeouts are a common issue.

jkelling commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -142,62 +125,264 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
-            return build(self, start_idx, end_idx, tmp_maxes, tmp_mins, _median, _compact);
+            mins[d] = fixed_val;
+            maxes[d] = fixed_val;
+            continue;
```
Ahh, right, they are reused later when climbing back up from the `_less` branch. I will have to copy. Recursion inside `split_and_partition` would be very awkward.

jkelling replied:
Added copy in case `mins`, `maxes` need to be modified in the `_compact=false` case.

jkelling commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -142,62 +125,264 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+{
+    std::vector<ckdtreenode>& buf = buflist->front();
+
+    /* put a new node into the node stack */
+    const ckdtree_intp_t node_index = buf.size();
+    ckdtree_intp_t _less;
+    buf.emplace_back();
+    ckdtreenode *node = &buf.back();
+    init_ckdtreenode(*node, start_idx, end_idx);
+
+    if (end_idx-start_idx <= self->leafsize) {
+        /* below brute force limit, return leafnode */
+        node->split_dim = -1;
+        return node_index;
+    }
```
A heuristic will always have cases where it does not make the optimal choice. This decision is also only relevant for small trees which will for one usually be run with one worker and where the performance does not matter that much (the docstring of `KDTree` states somewhere, that most time is usually spend in querying). In the case of an extremely imbalanced tree, where this case could be hit on the short branch of a large tree, this also does not matter, because the runtime will be dominated by the worker processing the deep branch.

The user can control this behavior using the `workers` parameter, if they care. However, if someone is building many small trees, they are better off using `workers=1` and build the trees in parallel in any case.

This suggestion optimizes for an irrelevant case in an inflexible way, which is why I oppose it.

peterbell10 replied:
>This suggestion optimizes for an irrelevant case in an inflexible way, which is why I oppose it.

It also makes the code simpler, so even if you don't care about efficiency it's still cleaner code.

jkelling replied:
The early check of `workers` does make it simpler, but I prefer not to add the small subtree part.

jkelling replied:
Added early `workers==1` check and call to sequential version of `build()` to eliminate `build_recurse()`.

jkelling commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -142,62 +125,264 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+          int workers)
+    {
+        const int subworkers = workers/2;
+        if(subworkers > 0) {
```
Done.

sturlamolden commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -142,62 +127,259 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+        worker.wait();
+        buflist->splice(buflist->end(), worker.buflist());
+        node->_greater = _greater;
+
```
```node->split_dim``` and ```node->split``` are not filled in

jkelling replied:
The are set by `split_and_partition`, here:
https://github.com/scipy/scipy/blob/017f9a329196a0a4026805607c5e2230494e8121/scipy/spatial/ckdtree/src/build.cxx#L146-L147

sturlamolden commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -142,62 +127,259 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+        node->_less = _less;
+        node->less = buf.data() + node->_less;
+        node->_greater = _greater;
+        node->greater = buf.data() + node->_greater;
```
```node->split_dim``` and ```node->split``` are not filled in

sturlamolden replied:
They are exposed to Python, so they must be saved, although they are no longer needed for the kd-tree.

jkelling replied:
Both `build` functions call `split_and_partition`, which takes care of this.

sturlamolden replied:
Ok

sturlamolden commented:
This is looking much better now.

peterbell10 commented on benchmarks/benchmarks/spatial.py:
```diff
@@ -34,15 +35,15 @@ def setup(self, mnr, cls_name):
         """
         m, n, r = mnr
         if cls_name == 'cKDTree_flat':
-            self.T = self.cls(self.data, leafsize=n)
+            self.T = self.cls(self.data, leafsize=n, workers=workers)
         else:
             self.cls(self.data)
```
This doesn't benchmark `workers`. `cKDTree_flat` isn't one of the options, so I don't know why that branch is there and you've edited the wrong side.

Also note that benchmarks need to work with older versions of scipy so we can compare performance over time. So, in the setup you should detect if there is a `workers` argument, and only add it if there is.

```
try:
    self.cls(np.ones((1, 3)), workers=1)
    self.workers_arg = dict(workers=1)
except:
    self.workers_arg = {}
```

jkelling replied:
Adressed in https://github.com/scipy/scipy/pull/14481/commits/d965a626aae7f2f90f4367abd753b3d52eb4061d .

jkelling replied:
@peterbell10 can you please check this? CI (specifically Lint) passes now. I added a function to do the filtering. in these two places.

peterbell10 commented on benchmarks/benchmarks/spatial.py:
```diff
@@ -72,14 +73,14 @@ def setup(self, mnr, balanced, order, radius):
-    def time_build(self, mnr, balanced, order):
-        cKDTree(self.data.get(order), balanced_tree=balanced)
+    def time_build(self, mnr, balanced, order, workers):
+        cKDTree(self.data.get(order), balanced_tree=balanced, workers=workers)
```
Same issue here.

peterbell10 commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -12,42 +12,27 @@
 #include <typeinfo>
 #include <stdexcept>
 #include <ios>
+#include <future>
+#include <atomic>
+#include <list>
+#include <memory>
```
Some of these are unused, I think.
```suggestion
#include <future>
#include <list>
```

peterbell10 commented on scipy/spatial/ckdtree/src/build.cxx:
```diff
@@ -142,62 +127,259 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+        std::copy_n(maxes, maxes_.size(), maxes_.begin());
+
+        future_ = std::async(std::launch::async, [&]() {
+                buflist_.emplace_back();
+                buflist_.back().reserve((end_idx-start_idx)/self->leafsize/workers);
```
Why call `emplace_back` and `reserve` each time you call build, instead of just calling it inside of `build`?

jkelling replied:
In a way this is left over from when there was only one build for both the parallel and the sequential recursion. Now it basically does not matter, both places are equivalent in terms of code sequence. However, as `buflist_` is a member of `build_worker`, so it seems sensible to me to initialized it here.

peterbell10 replied:
> Now it basically does not matter, both places are equivalent in terms of code sequence. 

It is equivalent yes, but this way it's copied to two places that must be kept in sync vs. only one place at the top of `build`.

jkelling replied:
Fixed in https://github.com/scipy/scipy/pull/14481/commits/1776bb188f00738d69f0f727f1ee54eac9c86806 .

peterbell10 reviewed (commented):
Looks good to me as well. Some minor issues and also, this really needs tests.

jkelling commented:
> Also, it would be nice to have some guidelines on choosing number workers in the doc string if we already have these info:
> 
>     * Do we use more resources as workers increase. How does this scale?

The added resources are marginal for large trees, the biggest part should be the thread itself. In terms of data held by the code explicitly there are only the added `std::vector` and `std::list` heads per worker. Going from on to two workers will double the memory required by the tree temporarily, because the nodes are copied form the temporary chunks to the final list after the build. This does not go up further with larger numbers of workers.
 
>     * Should the number of workers match the number of cores available in the system?

The parallel scaling is not very good, using more than ~8 workers will not give your useful speedup. It doe not slow it though either, just the parallel efficiency tanks.

>     * What happens if one calls threaded tree build from multiple Python threads? Mixing different threading systems used to be fragile -- over-subscription can negatively affect the performance.

Oversubscription will affect performance negatively, there is nothing to be done about that. There should be no problem staring C++11 threads (pthreads on Linux) from any python thread.

I think there caveats here which are not general to any threaded implementation, including the already parallel `query()`.

peterbell10 commented:
>Sometimes people use a convention where non-zero value of workers (!= 0) triggers the parallel code branch.
This way workers == 1 represents a single worker running through the parallel code branch, and workers == 0 represents a single worker running through the sequential code branch. All configurations are represented.

Is there any practical reason you'd want to run the parallel version with only 1 thread? It's just extra overhead. Also, this isn't the behaviour of any other parallel operations in SciPy.

sturlamolden commented:
Presumably, in the context of Python, ```workers=0``` would mean to retain the GIL during the operation.

sturlamolden commented:
> The parallel scaling is not very good, using more than ~8 workers will not give your useful speedup. It doe not slow it though either, just the parallel efficiency tanks.

At each node, we do O(N) operations (N being the number of data assigned to the node) before workers are spawned, e.g. ```std::nth_element```. Diminishing returns are therefore expected. A parallel introselect might help. To get good scalability we would need to ensure that any O(N) operation is done in parallel, including things like finding min and max values. Is there a parallel STL implementation we could use?

jkelling commented:
> It is a typical fork-join scalability problem, when we do sequential O(N) before the fork or after the join.

I did mention this in the description of this PR: parallel STL was introduced in C++17. Before starting this PR I checked SciPy's contribution guide and found, that only C++11 is allowed. I also found a CI check somewhere using GCC 4.8, which does support C++11, except `std::regex`.

So: In theory yes, in practice no.

rainwoodman commented:
> any other parallel operations in SciPy.
which probably should have followed the same ==0 and !=0 convention.

Other than the 'completeness', I wonder if there are still use cases that forbid the use C++ threading but still desire to use scipy's cKdtree. I don't know what these are, perhaps imagine some real-time systems, or a legacy platform that doesn't support c++'s threads. These use cases will have to look elsewhere without the ==0 special case.

jkelling commented:
> Other than the 'completeness', I wonder if there are still use cases that forbid the use C++ threading but still desire to use scipy's cKdtree. I don't know what these are, perhaps imagine some real-time systems, or a legacy platform that doesn't support c++'s threads. These use cases will have to look elsewhere without the ==0 special case.

This case would require a compile-time switch: If C++ threads are not supported, the code would just not compile, i.e. scipy would not build on that platform, checking `workers==0` at runtime does not help with that. Scipy does [allow C++11](https://docs.scipy.org/doc/scipy/reference/dev/toolchain.html), albeit only recently, which includes threads. Using standard library features is more portable than using an external, possible OS-dependent, library.

peterbell10 commented:
>Other than the 'completeness', I wonder if there are still use cases that forbid the use C++ threading but still desire to use scipy's cKdtree. I don't know what these are, perhaps imagine some real-time systems, or a legacy platform that doesn't support c++'s threads. These use cases will have to look elsewhere without the ==0 special case.

We don't support such a platform, but even if that were true this is a separate discussion. The current situation is that `workers=1` runs single threaded code, and doesn't have any synchronization overhead. Your proposal would mean that `workers=1` should actually use multithreading primitives and break on this hypothetical platform.

sturlamolden commented:
> Other than the 'completeness', I wonder if there are still use cases that forbid the use C++ threading but still desire to use scipy's cKdtree. I don't know what these are, perhaps imagine some real-time systems, or a legacy platform that doesn't support c++'s threads. These use cases will have to look elsewhere without the ==0 special case.

Real-time systems cannot use Python ```threading.Thread``` due to cooperative multitasking (because of the GIL). Real-time systems require preemptive multitasking. It might actually be a bigger problem that we use Python threads in the query methods. We should rather consider to use C++11 threads everywhere in cKDTree and get rid of the Python threading.

sturlamolden commented:
Can you please stop the excessive force pushing? 

Just do a final squash at the end, when you are done.

jkelling commented:
> Can you please stop the excessive force pushing?
> 
> Just do a final squash at the end, when you are done.

Ok, sorry. I am currently just struggelling with the python linter regarding the asv benchmarks. I cannot run the tests on my system because of https://github.com/scipy/scipy/issues/14501 . There is no use to reviewing at this point, as the CI fails (specifically lint) fails, but I can squash just before requesting a new review.

tupui commented:
What @sturlamolden wanted to say, is that we usually don't rebase force push at all. The simple reason is that it breaks the review history and makes reviewing harder.

At the end, before merging only, we either ask to clean the branch or do a squash.

In any case, thanks @jkelling for this work!

jkelling commented:
> At each node, we do O(N) operations (N being the number of data assigned to the node) before workers are spawned, e.g. `std::nth_element`. Diminishing returns are therefore expected. A parallel introselect might help. To get good scalability we would need to ensure that any O(N) operation is done in parallel, including things like finding min and max values. Is there a parallel STL implementation we could use?

@sturlamolden I Answered in one way above, still I tested this with gcc 10: It turns out, that `std::nth_element` and `std::partition` are not implemented in parallel in this implementation of the STL. I.e. adding calling with `par` or `par_unseq`, does not lead to parallel execution, while the gcc 10 STL implementation can execute some other algorithms in parallel.

There would be some other problem though: In C++17 there is not way to limit the number of threads used by these parallel algorithms. There are proposals about this, but nothing in the standard yet. Thus, we cold not impose the limit `workers`.

sturlamolden commented:
> What @sturlamolden wanted to say, is that we usually don't rebase force push at all. The simple reason is that it breaks the review history and makes reviewing harder.
> 
> At the end, before merging only, we either ask to clean the branch or do a squash.

Yes, that is the polite formulation, sorry for my rude English.

Being Scandinavian I have the same disability as Linus. We talk in a particular way that is easily perceived as offensive or at least impolite.

tupui commented:
> Yes, that is the polite formulation, sorry for my rude English.
> 
> Being Scandinavian I have the same disability as Linus. We talk in a particular way that is easily perceived as offensive or at least impolite.

No problem, thanks the clarification. Since we just have the writing and most of the time don't know each others, I know how it can be difficult. Don't be afraid of using a smiley or two 😃😅 Some people hate this, but it does help to convey emotions easier in a written conversation without writing much.

ev-br commented:
> we usually don't rebase force push at all. 

Some of do. And there is nothing wrong with force-pushes per se

tupui commented:
> > we usually don't rebase force push at all.
> 
> Some of us do. And there is nothing wrong with force-pushes per se

True, true, of course the story is more complicated. It depends on how the review goes. If a reviewer left some comment on a commit and then there is a squash (and then force push). The links are lost and you have GitHub telling you that it cannot find the link. But if you force push commits that are strictly after previous review comments, then all good.

jkelling commented on scipy/spatial/ckdtree.pyx:
```diff
@@ -602,6 +606,18 @@ cdef class cKDTree:
+        if workers == -1:
+            build_workers = os.cpu_count()
+            if build_workers is None:
+                raise NotImplementedError(
+                    "Cannot determine the number of cpus using os.cpu_count(), "
+                    "cannot use -1 for the number of workers"
+                )
```
I copied this from here:
https://github.com/scipy/scipy/blob/bdee8002e22a17ec8b7cf6324787428be6c1ab5a/scipy/spatial/ckdtree.pyx#L383-L406

Maybe the idea was that, if `os.cpu_count()` does not work it is not implemented on the present platform. I do, however agree, that from the perspective of cKDTree this is more of a `RuntimeError`.

Using `RuntimeError` in this PR would make the behavior inconsistent with `query()` though. So:
1. Do you want the inconsistency or
2. should we stick to NotImplementedError or
3. should this other place be changed, too?

tupui replied:
We also have a logic like this in scipy.stats.qmc
https://github.com/scipy/scipy/blob/33a25aea48bea84fee7861355cbccf36a3fc90a6/scipy/stats/_qmc.py#L1721-L1748

jkelling replied:
I think changing this should not be part of this PR. I will mark this as resolved.

peterbell10 reviewed (changes requested):
This still needs tests.

jkelling commented:
> This still needs tests.

Unfortunately, GitHub doe snot show me what your are referring to. There is "View changes" Button, but this only switches to the "Files changed" tab which displays the whole diff, but doe not show any comments.

Are you referring to the function I added in [the asv benchmark](https://github.com/scipy/scipy/pull/14481/files#diff-5c9e7e8948e7ff6efc82ab31a7edc2f9cc09abcd130808520be342f4afa76668R19-R38)?

peterbell10 commented:
>Unfortunately, GitHub doe snot show me what your are referring to. 

It's a top level comment. I'm referring to the whole PR. Nothing here is tested because the default workers is 1.

You need to add a parallel build test in `test_kdtree.py`:
https://github.com/scipy/scipy/blob/master/scipy/spatial/tests/test_kdtree.py

jkelling commented:
> You need to add a parallel build test in `test_kdtree.py`:
> https://github.com/scipy/scipy/blob/master/scipy/spatial/tests/test_kdtree.py

Ok, thanks.

then it would probably be good to have most of the coverage provided the existing consistency tests for the parallel case. Would it be Ok to make those tests parallel-"only"? Having a parallel build would still cover most of the non-parallel code. Covering the in-place bit for `workers=1` by only one other test should be enough as the cases only diverge in [these lines](https://github.com/scipy/scipy/blob/17b6d03fb6794fe6a827f60a7f565a53b94269b6/scipy/spatial/ckdtree/src/build.cxx#L329-L334).

How many threads can the CI take? More worker would be better at catching race conditions, but `workers=2` would be enough to go down the code path. There is no intentionally shared data, so there should be no race conditions---only an indexing bug could cause data races, but this type of bug can also be caught with a sequential test (or few threads) without forcing a data race to occur.

peterbell10 commented:
>then it would probably be good to have most of the coverage provided the existing consistency tests for the parallel case. Would it be Ok to make those tests parallel-"only"? 

Ideally you would test both. I would suggest doing a few large tree builds with `workers=1` and `workers=2` and check they build the same tree.

>How many threads can the CI take? More worker would be better at catching race conditions, but workers=2 would be enough to go down the code path.

CI workers only have 2 threads, so I would stick to that. This is more to catch blatant logic errors.

jkelling commented:
The tests are in now. Uncovered one bug in the code (required the lower-than-workers CPU count in the CI to show).

I rebased onto master. Because files were had been renamed on master, I squashed all commits before the one adding the test to make the rebase simpler. Discussions on those changes were resolved.

If you approve, I would like to squash this branch before it is merged.

jkelling commented:
The CI failures are unrelated to the changes in this PR.

jkelling commented:
Anyone available to review the tests added by e5aadc79f3950269f1fd74c497b88e1fe58d1f12 ?

sturlamolden commented:
> Anyone available to review the tests added by [e5aadc7](https://github.com/scipy/scipy/commit/e5aadc79f3950269f1fd74c497b88e1fe58d1f12) ?

LGTM, but personally I would stress-test with an even higher number of workers.

jkelling commented:
> LGTM, but personally I would stress-test with an even higher number of workers.

Thanks.

For a stress-test, if we went `(2,3,6,11,16)` we would reach the largest relevant numbers (it is realistic to run more workers, but parallel efficiency tanks...). I guess we should keep the number of cases for this parameter small because the test also goes through (1...5) dimensional trees, so this suggestion would already lead to 25 tree builds.

Also: we need the lower numbers too, because some race conditions will only show with actual concurrency and with a large number if threads ensuing preemptive multi-tasking may just "fix" some of them.
